File: typescript/src/sort/quick.ts
function qs(arr: number[], lo: number, hi: number): void {
  if (lo >= hi) {
    return;
  }

  let pivotIdx = partition(arr, lo, hi);

  qs(arr, lo, pivotIdx - 1);
  qs(arr, pivotIdx + 1, hi);
}

function partition(arr: number[], lo: number, hi: number): number {
  const pivot = arr[hi];

  let idx = lo - 1;

  for (let i = lo; i < hi; ++i) {
    if (arr[i] <= pivot) {
      idx++;

      const temp = arr[i];
      arr[i] = arr[idx];
      arr[idx] = temp;
    }
  }

  idx++;

  arr[hi] = arr[idx];
  arr[idx] = pivot;

  return idx;
}

const input = [9, 1, 4, 8, 3, 6];

// input === [1, 3, 4, 6, 8, 9];

File: rust/sort/src/quick.rs
pub fn partition(arr: &mut Vec<usize>, low: usize, high: usize) -> usize {
    let pivot = arr[high];

    let mut idx = low;

    for i in low..high {
        if arr[i] <= pivot {
            arr.swap(i, idx);
            idx += 1;
        }
    }

    arr.swap(high, idx);
    arr[idx] = pivot;

    idx
}

pub fn qs(arr: &mut Vec<usize>, low: usize, high: usize) {
    if low >= high {
        return;
    }

    let pivot_idx = partition(arr, low, high);

    if let Some(pivot) = pivot_idx.checked_sub(1) {
        qs(arr, low, pivot);
    }

    if let Some(pivot) = pivot_idx.checked_add(1) {
        qs(arr, pivot, high);
    }
}

#[cfg(test)]
mod tests {
    use super::qs;

    #[test]
    fn quick_sort() {
        let mut arr = vec![8, 1, 7, 3, 5, 2];

        let len: usize = arr.len() - 1;

        qs(&mut arr, 0, len);

        assert_eq!(vec![1, 2, 3, 5, 7, 8], arr);
    }
}
